client/transaction-pool/src/lib.rs

```diff
@@ -563,167 +585,296 @@ async fn prune_known_txs_for_block<Block: BlockT, Api: graph::ChainApi<Block = B
+where
+	Block: BlockT,
+	PoolApi: 'static + graph::ChainApi<Block = Block>,
+{
+	async fn maintain(&self, event: ChainEvent<Self::Block>) {
+		let prev_finalized_block = self.enactment_state.lock().recent_finalized_block;
+		let compute_tree_route = |from, to| -> Result<TreeRoute<Block>, String> {
+			match self.api.tree_route(from, to) {
+				Ok(tree_route) => Ok(tree_route),
+				Err(e) =>
+					return Err(format!(
+						"Error [{e}] occurred while computing tree_route from {from:?} to \
+					 previously finalized: {to:?}"
+					)),
+			}
+		};
+
+		let result = self
+			.enactment_state
+			.lock()
+			.update_and_check_if_new_enactment_is_valid(&event, &compute_tree_route);
+		match result {
+			Err(msg) => {
+				log::warn!(target: "txpool", "{msg}");
+				return
 			},
-			ChainEvent::Finalized { hash, tree_route } => {
-				let pool = self.pool.clone();
-				async move {
-					for hash in tree_route.iter().chain(&[hash]) {
-						if let Err(e) = pool.validated_pool().on_block_finalized(*hash).await {
-							log::warn!(
-								target: "txpool",
-								"Error [{}] occurred while attempting to notify watchers of finalization {}",
-								e, hash
-							)
-						}
-					}
-				}
-				.boxed()
+			Ok(None) => {},
+			Ok(Some(tree_route)) => {
+				self.handle_enactment(tree_route).await;
 			},
+		};
+
+		if let ChainEvent::Finalized { hash, tree_route } = event {
+			log::trace!(target:"txpool", "on-finalized enacted: {tree_route:?}, previously finalized: \
+						{prev_finalized_block:?}");
+			for hash in tree_route.iter().chain(std::iter::once(&hash)) {
+				if let Err(e) = self.pool.validated_pool().on_block_finalized(*hash).await {
+					log::warn!(
+						target: "txpool",
+						"Error [{}] occurred while attempting to notify watchers about finalization {}",
+						hash, e
+					)
+				}
+			}
+		}
+	}
+}
+
+/// Helper struct for deciding if core part of maintenance procedure shall be executed.
+/// It is publicly exposed only for tests purposes (due to cyclic deps sc_transaction_pool and
+/// substrate_test_runtime_transaction_pool).
+///
+/// For the following chain:
+///   B1-C1-D1-E1
+///  /
+/// A
+///  \
+///   B2-C2-D2-E2
+///
+/// Some scenarios and expected behavior for seqeunce of `NewBestBlock` (nbb) and `Finalized` (f)
+/// events: `nbb(C1)`, `f(C1)` -> false (handle_enactment was already performed in `nbb(C1))`
+/// `f(C1)`, `nbb(C1)` -> false (handle_enactment was already performed in `f(C1))`
+///
+/// `f(C1)`, `nbb(D2)` -> false (handle_enactment was already performed in `f(C1)`, we should not
+/// retract finalized block)
+/// `f(C1)`, `f(C2)`, `nbb(C1)` -> false
+/// `nbb(C1)`, `nbb(C2)` -> true (switching fork is OK)
+/// `nbb(B1)`, `nbb(B2)` -> true
+/// `nbb(B1)`, `nbb(C1)`, `f(C1)` -> false (handle_enactment was already performed in `nbb(B1)`
+/// `nbb(C1)`, `f(B1)` -> false (handle_enactment was already performed in `nbb(B2)`
+struct EnactmentState<Block>
```

Comment:
I think putting this and the tests for this struct into its own file would be reasonable.

Reply:
[649bbf3](https://github.com/paritytech/substrate/pull/12305/commits/649bbf382cf336607aa088a995afdae3f1a444d9)